### PR TITLE
Add management cluster KAS network policy

### PR DESCRIPTION
**Pull request by enxebre**
**What this PR does / why we need it**:
As part of hardening CP workloads, this introduces a network policy that:

Selects pods excluding the ones having NeedManagementKASAccessLabel and specific operands.

It denies egress traffic to the management cluster clusterNetwork and to the KAS endpoints.

**Which issue(s) this PR fixes** *(optional, use `fixes #<issue_number>(, fixes #<issue_number>, ...)` format, where issue_number might be a GitHub issue, or a Jira story*:
Fixes #

**Checklist**
- [ ] Subject and description added to both, commit and PR.
- [ ] Relevant issues have been referenced.
- [ ] This change includes docs. 
- [ ] This change includes unit tests.

### CODE REVIEW

File: test/e2e/util/util.go [hunk 1]
```diff
@@ -727,6 +728,113 @@ func EnsureAllRoutesUseHCPRouter(t *testing.T, ctx context.Context, hostClient c
...
+		})
+
+		t.Run("EnsureEgressTrafficToManagementKAS", func(t *testing.T) {
+			command := []string{
+				"curl",
+				"--connect-timeout",
+				"2",
+				"-k",
+				// Default KAS advertised address.
+				"https://172.20.0.1:6443",
```
Comment (muraee) [inline test/e2e/util/util.go:762]: where did this IP come from? isn't better to use use the Kubernetes service DNs record `https://kubernetes.default.svc:443`?

File: test/e2e/util/util.go [hunk 2]
```diff
@@ -727,6 +728,113 @@ func EnsureAllRoutesUseHCPRouter(t *testing.T, ctx context.Context, hostClient c
...
+	podExecuter := PodExecOptions{
+		StreamOptions: StreamOptions{
+			IOStreams: genericclioptions.IOStreams{
+				Out:    stdOut,
+				ErrOut: stdErr,
+			},
+		},
+		Command:   command,
+		Namespace: namespace,
+		PodName:   podList.Items[0].Name,
```
Comment (muraee) [inline test/e2e/util/util.go:830]: you would need to pass `ContainerName` as well.

File: test/e2e/util/util.go (hunk 1 above)
Comment (enxebre) [reply test/e2e/util/util.go:762]: This is our `DefaultAdvertiseAddress = "172.20.0.1"`.
I want to query the pod network directly to remove chances of Service network resolution hiding the real test.

**Comment by enxebre** [comment]
Last run failed because of https://github.com/openshift/hypershift/pull/2757

/test e2e-aws

**Comment by enxebre** [comment]
>        --- FAIL: TestNodePool/NodePool_Tests_Group/TestNTOMachineConfigGetsRolledOut (1490.10s)
        --- FAIL: TestNodePool/NodePool_Tests_Group/TestNodepoolMachineconfigGetsRolledout (1490.10s)

/test e2e-aws

File: hypershift-operator/controllers/hostedcluster/network_policies.go [hunk 3]
```diff
@@ -542,3 +566,124 @@ func reconcileSameNamespaceNetworkPolicy(policy *networkingv1.NetworkPolicy) err
...
+	// i.e. block access to management cluster KAS.
+	exceptions := append(kasCIDRs, clusterNetworks...)
+	policy.Spec.Egress = append(policy.Spec.Egress,
+		networkingv1.NetworkPolicyEgressRule{
+			To: []networkingv1.NetworkPolicyPeer{
+				{
+					IPBlock: &networkingv1.IPBlock{
+						CIDR:   "0.0.0.0/0",
+						Except: exceptions,
+					},
```
Comment (davidvossel) [inline hypershift-operator/controllers/hostedcluster/network_policies.go:620]: does ipv6 need to be accounted for here as well?

**Review by davidvossel — COMMENTED** [review]
One difference between the kubevirt lane and the aws lane is that the kubevirt lane's mgmt cluster is not a HCP. meaning our management cluster is a standard ipi install of aws. 

I'm still unable to definitively identify why the kubevirt lane is failing, but it's possible it's related to this difference... and if that's the case, then it's also possible that it's not just related to kubevirt, but it's instead a side effect of how the kubevirt lane is tested vs how the aws lane is tested.

**Pull request merged**